runtime/bin/directory_linux.cc

```diff
@@ -120,6 +121,11 @@ ListType DirectoryListingEntry::Next(DirectoryListing* listing) {
   // ports.
   errno = 0;
   dirent* entry = readdir(reinterpret_cast<DIR*>(lister_));
+  if (entry != NULL) {
+    const uint8_t* d_name = reinterpret_cast<uint8_t *>(entry->d_name);
+    if (!Utf8::IsValid(d_name, strlen(entry->d_name)))
+      return kListError;
```

Comment:
I wonder if we should instead just skip entries that aren't utf8 formatted. So instead of "return kListError", "return Next(listing)", and update the docs to indicate the behavior.

Reply:
I think this is reasonable, yes. These were the choices I've thought about a little bit:
1) Make it so that file / directory / ... names can be arbitrary sequences of bytes. Advantage is that it's fully general, e.g. one could then implement a filesystem explorer in Dart that allows the user to delete even a file with incorrect utf8 chars. But I suspect this is a very difficult change to get done. :-)
2) Coerce the names into valid utf8. This is relatively easy and has the advantage that in the hypothetical filesystem explorer, the user could see that there is a file or directory with a weird name. But some operations may not work, e.g., deleting may be hard to implement because the Dart program isn't able to tell the OS precisely which file to delete, and some other operations may be a bit quirky too.
3) Skip over files that aren't valid utf8. This means the user won't even see what's there - this can be puzzling e.g. they may look at a volume like a usb stick in their filesystem explorer and wonder why there's no space available even though there isn't a file to see.
4) Make the error easier to understand. The filesystem explorer may still crash when pointed at a directory, but if someone looks at the stack trace they may be able to understand why. And perhaps sometimes it's possible for the hypothetical filesystem excplorer to catch the FileSystemException and handle it somehow.
I don't think I have enough context to pick. 4) seemed like the least intrusive and leave the door open for any of the others later. :-) I'd be happy to attempt 2) or 3) if you like.